src/Features/Core/Portable/PullMemberUp/AbstractPullMemberUpRefactoringProvider.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp.Dialog;
+using Microsoft.CodeAnalysis.PullMemberUp.QuickAction;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal abstract partial class AbstractPullMemberUpRefactoringProvider : CodeRefactoringProvider
+    {
+        protected readonly IPullMemberUpOptionsService _pullMemberUpOptionsService;
+
+        internal AbstractPullMemberUpRefactoringProvider(IPullMemberUpOptionsService pullMemberUpService)
+        {
+            _pullMemberUpOptionsService = pullMemberUpService;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var userSelectedNode = root.FindNode(context.Span);
+
+            if(!IsUserSelectIdentifer(userSelectedNode, context))
+            {
+                return;
+            }
+
+            var userSelectNodeSymbol = semanticModel.GetDeclaredSymbol(userSelectedNode);
+            if (userSelectNodeSymbol == null)
+            {
+                return;
+            }
+
+            var allTargetClasses = FindAllTargetBaseClasses(userSelectNodeSymbol.ContainingType, context.Document.Project.Solution, context.CancellationToken);
+            var allTargetInterfaces = FindAllTargetInterfaces(userSelectNodeSymbol.ContainingType, context.Document.Project.Solution, context.CancellationToken);
+
+            if (allTargetInterfaces.Count == 0 && allTargetClasses.Count == 0)
+            {
+                return;
+            }
+
+            if (userSelectNodeSymbol.Kind == SymbolKind.Method ||
+                userSelectNodeSymbol.Kind == SymbolKind.Property ||
+                userSelectNodeSymbol.Kind == SymbolKind.Event)
+            {
+                await AddPullUpMemberToClassRefactoringViaQuickAction(allTargetClasses, userSelectedNode, userSelectNodeSymbol, context);
+                await AddPullUpMemberToInterfaceRefactoringViaQuickAction(allTargetInterfaces, userSelectedNode, userSelectNodeSymbol, context);
+                AddPullUpMemberRefactoringViaDialogBox(userSelectNodeSymbol, context, root, semanticModel);
```

Comment:
can i see a picture of this in action?

Reply:
order vairables from 'widest to narrowest'.  so context, semanticmodel, symbol, root.